torch/autograd/function.py

```diff
@@ -200,30 +186,43 @@ def once_differentiable(fn):
 
     @functools.wraps(fn)
     def wrapper(ctx, *args):
-        tensor_args = [arg.data if isinstance(arg, Variable) else arg
-                       for arg in args]
-        outputs = fn(ctx, *tensor_args)
-        # XXX: this is only an approximation of these flags - there's no way
+        with torch.no_grad():
+            outputs = fn(ctx, *args)
+
+        if not torch.is_grad_enabled():
+            return outputs
+
+        # If any of the inputs have requires_grad=True, we force the outputs
+        # to have requires_grad=True but point to a grad_fn which throws an
+        # error message during (double) back-propagation.
+        # XXX: this is only an approximation of requires_grad - there's no way
         # to figure out if fn didn't use ctx.saved_variables and as a result
         # some Variables might require grad, even if no args do.
         # Unfortunately, this leads to unexpected error messages ("no nodes
         # require computing gradients"), but I don't have a better idea.
         # These functions would raise an error in backward anyway.
-        requires_grad = any(arg.requires_grad if isinstance(arg, Variable) else False
+        requires_grad = any(isinstance(arg, Variable) and arg.requires_grad
                             for arg in args)
-        if not torch.is_grad_enabled():
-            def err_fn(*args):
-                return args
-        else:
-            err_fn = torch._C._functions.DelayedError(
-                b"trying to differentiate twice a function that was marked"
-                b"with @once_differentiable")
+        if not requires_grad:
+            return outputs
+
+        err_fn = torch._C._functions.DelayedError(
+            b"trying to differentiate twice a function that was marked"
+            b"with @once_differentiable")
+
         if not isinstance(outputs, tuple):
-            var = (Variable(outputs, requires_grad=requires_grad)
-                   if outputs is not None else None)
-            return err_fn(var)
-        return err_fn(*[Variable(o, requires_grad=requires_grad) if o is not None else None
-                      for o in outputs])
+            outputs = (outputs,)
+
+        # Create aliases of each output that has requires_grad=True. We need
+        # at least one of the inputs to err_fn to require grad so that the
+        # output will have a grad_fn.
+        def fake_requires_grad(var):
+            if var is not None:
+                var = var.detach()
+                var.requires_grad = True
```

Comment:
will they become leaves now? (and will accumulate grad)

Reply:
`fake_requires_grad` creates leaf variables, but the thing that is returned is:

 ```
err_fn(*[fake_requires_grad(v) for v in outputs])
```

The application of `err_fn` creates non-leaf Variables. Accumulate grad will never be called because the backward of `err_fn` throws an exception.